c_src/sasl_auth.c

```diff
@@ -134,24 +238,56 @@ static ERL_NIF_TERM sasl_cli_new(ErlNifEnv* env, int UNUSED(argc), const ERL_NIF
 
     state->mech_set = 0;
 
+    enif_self(env, &state->controlling_process);
+
+    state->controller_lock = enif_mutex_create("sasl_auth.controller_lock");
+
+    state->principal = copy_bin(principal);
```

Comment:
This fixes a thing we should have never been doing and could result in memory corruption (treating the argument as non-transient).